robot_calibration/include/robot_calibration/capture_features.h

```diff
@@ -0,0 +1,124 @@
+*
+* @return     string representing the absolute path to the file
+*/
+std::string get_absolute_directory(std::string local_dir);
+
+
+/**
+* @brief      Creates a map of the available feature finders
+*
+* @param[in]  nh        node handle reference.
+* @param[in]  finders   the location to store the map of feature finders
+*
+* @return     bool indicating whether the feature finders were loaded 
+*/
+bool get_feature_finders(ros::NodeHandle& nh, robot_calibration::FeatureFinderMap& finders);
+
+
+/**
+* @brief      Executes the steps required to automatically capture the feature from a bag of predetermined pose.
+*             This includes commanding the robot to the specified pose then capturing the feature. 
+*
+* @param[in]  nh                node handle reference.
+* @param[in]  chain_manager     object that handles the kinematic chains
+* @param[in]  finders           map of available feature finders
+* @param[in]  feature           feature to use for calibration
+* @param[in]  bag               rosbag object for storing calibration data
+*
+* @return     bool indicating whether the run was successful 
+*/
+bool run_automatic_capture(ros::NodeHandle& nh, 
+                           robot_calibration::ChainManager* chain_manager, 
+                           robot_calibration::FeatureFinderMap& finders,
+                           std::string& feature,
+                           rosbag::Bag& bag);
+
+
+/**
+* @brief      Executes the steps required to manually capture the feature from a bag of predetermined pose.
+*             This includes waiting for the user to move the robot then capturing the feature on their command. 
+*
+* @param[in]  nh                node handle reference.
+* @param[in]  chain_manager     object that handles the kinematic chains
+* @param[in]  finders           map of available feature finders
+* @param[in]  feature           feature to use for calibration
+* @param[in]  bag               rosbag object for storing calibration data
+*
+* @return     bool indicating whether the run was successful   
+*/
+bool run_manual_capture(ros::NodeHandle& nh, 
+                        robot_calibration::ChainManager* chain_manager, 
+                        robot_calibration::FeatureFinderMap& finders,
+                        std::string& feature,
+                        rosbag::Bag& bag);
+
+
+/**
+* @brief      Opens the file and creates a bag object.
+*
+* @param[in]  filename  absolute path to the file.
+* @param[in]  bag       the bag object reference
+*
+* @return     bool indicating whether openning the file was successful
+*/
+bool open_bag(std::string filename, rosbag::Bag& bag);
+
+
+/**
+* @brief      Creates the bag object.
+*
+* @param[in]  filename  absolute path to the file.
+* @param[in]  bag       the bag object reference
+*
+* @return     bool indicating whether creating the file was successful
+*/
+bool create_bag(std::string filename, rosbag::Bag& bag);
```

Comment:
Here is another case where bag is an output parameter and you could use optional (assuming you have c++17).  If you don't have c++17 then maybe consider returning a pair<bool, rosbag::Bag>?